### PR TITLE
test: add local lxd container e2e testing for Derper, Headscale and Tailscale

### PR DESCRIPTION
**Pull request by Deezzir**
# Overview

The PR adds minimal e2e testing automation for a local environment. `tox -e e2e` can now be used on locally built snaps to test basic functionality and configuration drift, enabling faster, more streamlined releases.

Can be a good starting point towards closing https://github.com/canonical/tailscale-snap/issues/130

## Notes

The Python test suite can also be used in CI with Azure credentials to set up and test the snaps, for example by pulling from `latest/edge`.

We could also try to switch from containers to VMs, if needed.

Currently blocked by https://github.com/canonical/derper-snap/pull/41

### CODE REVIEW

**Comment by H-M-Quang-Ngo** [comment]
Currently if making change to any of Tailscale, Headscale and Derper snaps, we have to validate the whole setup using the e2e guide in Tailscale snap repo. This practice lasts for a while, but perhaps a more robust way is to migrate all e2e testing stuff to a centralised repo for the e2e testing, since the logic isn't specific to any. Wdyt?

I think, we can even take a further step to make it a CI action that can be used in all 3 snaps. So for PR in a repo, it will build corresponding snap artifact and install from store the other two (parameters can be supported to select the snap channel maybe?), then do all the provision as per this PR. 

Btw, thanks for the great effort on this!

File: tests/e2e/test_snap.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    )
+
+
+def _restore_derper_login(project: str, private_dir: Path) -> None:
+    enroll_with_key(project, "derper", "derper", private_dir, "tag:derper")
+    wait_tailscale(project, "derper")
+    _restart_derper(project)
+
+
+def test_custom_derp(e2e: E2EContext) -> None:
```
Comment (jneo8) [inline tests/e2e/test_snap.py:102]: nitpick:

I would like to see some doc-string to explain each test method's goal.

File: tests/e2e/run.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (jneo8) [inline tests/e2e/run.sh]: suggestion:

It will be better if we focus to use either script or python.
For example, the logic of check files and using terraform could be all move to python part.

(I think python maybe more make sense here since it's better for expansion)

**Review by jneo8 — CHANGES_REQUESTED** [review]
I really like the idea of small local e2e testing, the direction is very good.
Some suggestions on the implementation.

**Comment by Deezzir** [comment]
> Currently if making change to any of Tailscale, Headscale and Derper snaps, we have to validate the whole setup using the e2e guide in Tailscale snap repo. This practice lasts for a while, but perhaps a more robust way is to migrate all e2e testing stuff to a centralised repo for the e2e testing, since the logic isn't specific to any. Wdyt?

> I think, we can even take a further step to make it a CI action that can be used in all 3 snaps. So for PR in a repo, it will build corresponding snap artifact and install from store the other two (parameters can be supported to select the snap channel maybe?), then do all the provision as per this PR.

@H-M-Quang-Ngo @jneo8  I was thinking about that and honestly wanted to bring that on one of the standups, so we can discuss. I had two approaches in mind:

1. Replicate this e2e CI test for all snaps in question and, during the testing, build the current snap and pull the other two from `latest/edge`.
2. As proposed, we can move the CI to our automation repo and build all three and test there. 

I think that's what @H-M-Quang-Ngo is proposing; should we choose either or both?

WDYT?

**Comment by jneo8** [comment]
Hi @Deezzir 

I more prefer a simple and native way.
How about a shared github workflow/action? We can put it in one of repo and others import it.

**Comment by H-M-Quang-Ngo** [comment]
Yea, I think it's more convenient to maintain a single action that we can call in the workflows of all snaps. 

It will do all the required provision (as this PR implements), and, also run the suite of integration testing based on that.

File: tests/e2e/test_snap.py (hunk 1 above)
Comment (Deezzir) [reply tests/e2e/test_snap.py:102]: Addressed in a50e758e41c245da8d3523e5c0c108a73189f676

File: tests/e2e/run.sh (hunk 2 above)
Comment (Deezzir) [reply tests/e2e/run.sh]: Addressed in 8ac960f54cffc07263462041780f9782bdc0fb7e

**Comment by Deezzir** [comment]
@H-M-Quang-Ngo @jneo8 Please review the latest changes. I've removed `run.sh`, and everything is now invoked by the `conftest.py` fixtures. Also, added a CI workflow that can be reused in the future. Question: are we planning to also copy the `tests/e2e` for each target snap, or what would be the approach?

Also, @jneo8 please take a look https://github.com/canonical/tailscale-snap/actions/runs/31074597136, need a whitelist entry.

I guess we can use Renovate to pull `tests/e2e` and `.github/workflows/e2e.yaml` for Derper and Headscale from this repo?

**Comment by jneo8** [comment]
> Question: are we planning to also copy the tests/e2e for each target snap, or what would be the approach?

I am more prefer something base on git. How about our previous practice in [ovn-exporter](https://github.com/canonical/ovn-exporter)? Which base on submodules + softlink. 


> Also, @jneo8 please take a look https://github.com/canonical/tailscale-snap/actions/runs/31074597136, need a whitelist entry.

Added. Please try again. 

> I guess we can use Renovate to pull tests/e2e and .github/workflows/e2e.yaml for Derper and Headscale from this repo?

Renovate to auto-generate the PR to update the action version sound good to me.

**Comment by Deezzir** [comment]
@jneo8 The workflow runs, and it is expected to fail until https://github.com/canonical/derper-snap/pull/41 is merged, thanks

**Review by jneo8 — DISMISSED** [review]

**Comment by Deezzir** [comment]
@jneo8 @H-M-Quang-Ngo The tests now pass with the new derper rev 84

**Comment by jneo8** [comment]
@chanchiwai-ray Please help to review

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+name: End-to-end tests
+
+on:
+  workflow_call:
+  workflow_dispatch:
```
Comment (H-M-Quang-Ngo) [inline .github/workflows/e2e.yaml]: For workflow dispatch, I think it would be easier if we can give the input for the snap channel, at least for headscale (derper only has `latest/edge`).

Headscale is volatile between minor versions, so in case the breaking change is from a newer headscale (likely `latest/edge`) and fails the default test, IMO it would still be ok to release tailscale if the test passes with headscale on a more stable channel.

**Review by H-M-Quang-Ngo — APPROVED** [review]
It looks good to me, I just leave a small comment, but maybe it can be resolved in another PR.

Also it's for testing so I didn't go to very deep details.

Would like to try locally if having some times later.

**Review by chanchiwai-ray — APPROVED** [review]
